nav2_world_model/src/costmap_representation.cpp

```diff
@@ -0,0 +1,188 @@
+  //              to use the node passed to the constructor.
+  tf_node_ = rclcpp::Node::make_shared(name + "_tf_Node",
+    "",
+    rclcpp::contexts::default_context::get_global_default_context(),
+    std::vector<std::string>(),
+    std::vector<rclcpp::Parameter>(),
+    false,  // ignore global parameters, so this node doesn't get renamed
+    false,
+    false);
+
+  tfBuffer_ = std::make_shared<tf2_ros::Buffer>(tf_node_->get_clock()),
+  tfListener_ = std::make_shared<tf2_ros::TransformListener>(*tfBuffer_, tf_node_, true),
+  costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>(name_, *tfBuffer_);
+  costmap_ = costmap_ros_->getCostmap();
+  executor.add_node(costmap_ros_);
+}
+
+GetCostmap::Response
+CostmapRepresentation::getCostmap(const GetCostmap::Request & /*request*/)
+{
+  GetCostmap::Response response;
+  response.map.metadata.size_x = costmap_->getSizeInCellsX();
+  response.map.metadata.size_y = costmap_->getSizeInCellsY();
+  response.map.metadata.resolution = costmap_->getResolution();
+  response.map.metadata.layer = "Master";
+  response.map.metadata.map_load_time = costmap_ros_->now();
+  response.map.metadata.update_time = costmap_ros_->now();
+
+  tf2::Quaternion quaternion;
+  // TODO(bpwilcox): Grab correct orientation information
+  quaternion.setRPY(0.0, 0.0, 0.0);  // set roll, pitch, yaw
+  response.map.metadata.origin.position.x = costmap_->getOriginX();
+  response.map.metadata.origin.position.y = costmap_->getOriginY();
+  response.map.metadata.origin.position.z = 0.0;
+  response.map.metadata.origin.orientation = tf2::toMsg(quaternion);
+
+  response.map.header.stamp = costmap_ros_->now();
+  response.map.header.frame_id = frame_id_;
+
+  unsigned char * data = costmap_->getCharMap();
+  auto data_length = response.map.metadata.size_x * response.map.metadata.size_y;
+  response.map.data.resize(data_length);
+  response.map.data.assign(data, data + data_length);
+
+  return response;
+}
+
+ProcessRegion::Response
+CostmapRepresentation::confirmFreeSpace(const ProcessRegion::Request & request)
+{
+  ProcessRegion::Response response;
+  response.result = checkIfFree(request);
+  return response;
+}
+
+bool CostmapRepresentation::checkIfFree(const ProcessRegion::Request & request)
+{
+  // Define the vertices of a rectangle
+  std::vector<MapLocation> rectangle;
+  if (!generateRectangleVertices(request, rectangle)) {
+    RCLCPP_ERROR(node_->get_logger(),
+      "Could not generate rectangle vertices. Checking if region is free failed");
+    return false;
+  }
+
+  // Get all the cell locations inside the region
+  std::vector<MapLocation> polygon_cells;
+  costmap_->convexFillCells(rectangle, polygon_cells);
+
+  if (polygon_cells.empty()) {
+    RCLCPP_ERROR(node_->get_logger(),
+      "Could not generate rectangle cells. Checking if region is free failed");
+    return false;
+  }
+
+  // Check if there is at least one cell not free and send to visualizer for display
```

Comment:
Not sure I understand the issue, this is similar to how costmap_2d and other nav2 modules publish messages internally for visualization in RVIZ.

This is the robot's local world model, it's own point of view. Once we add more robots, each will have it's own, and then we'll have to introduce namespaces. 

At some point, it will be beneficial to have a shared global world model, but that will have a different architecture than `nav2_world_model`.